### ISSUE TITLE
Inplace aten::append is not converted by remove_mutation pass.

### ISSUE THREAD

**Issue by BowenBao** [pytorch/pytorch#39362]
# Description
aten::append is inplace and is not converted by remove_mutation pass.
Moreover, when used inside a loop, the output of aten::append is not captured in the loop output.

As a side note, it is not clear what should be captured as loop output as it is not mentioned https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#loops.
From looking at the `If` definition above, I'm assuming that loop would have similar definition for its output as `the outputs are the new values of variables in the outer block whose values were altered in an if-statement`. If the assumption is correct, then aten::append output should also be part of loop output.

# Code to repro
```python
import torch

class AppendModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(2, 2)
    def forward(self, x):
        res = []
        for i in range(x.size(0)):
            res.append(self.linear(x[i]))
        return torch.cat(res, dim=0)

m = torch.jit.script(AppendModel())
m.eval()

m._c.dump()

# freezing

m_freezed = torch._C._freeze_module(m._c)

m_freezed.dump()

# remove mutation
graph = m_freezed._get_method('forward').graph
torch._C._jit_pass_remove_mutation(graph)
print(graph)

"""
Output graph. Note that output(%10) of aten::append is not captured in the loop output.
And the aten::cat node outside of loop is using the original list value(%res.1) as input.

graph(%self : __torch__.___torch_mangle_0.AppendModel,
      %x.1 : Tensor):
  %27 : Float(2:1, 2:2) = prim::Constant[value=-0.1994  0.3128 -0.4812  0.3948 [ CPUFloatType{2,2} ]]()
  %13 : int = prim::Constant[value=2]() # /home/bowbao/repos/pytorch/torch/nn/functional.py:1672:22
  %12 : int = prim::Constant[value=1]()
  %3 : int = prim::Constant[value=0]() # repro_append.py:10:30
  %2 : bool = prim::Constant[value=1]() # repro_append.py:10:8
  %self.linear.bias : Float(2:1) = prim::Constant[value=-0.4483  0.6543 [ CPUFloatType{2} ]]()
  %res.1 : Tensor[] = prim::ListConstruct()
  %5 : int = aten::size(%x.1, %3) # repro_append.py:10:23
   = prim::Loop(%5, %2) # repro_append.py:10:8
    block0(%i.1 : int):
      %8 : Tensor = aten::select(%x.1, %3, %i.1) # repro_append.py:11:35
      %16 : int = aten::dim(%8) # /home/bowbao/repos/pytorch/torch/nn/functional.py:1672:7
      %17 : bool = aten::eq(%16, %13) # /home/bowbao/repos/pytorch/torch/nn/functional.py:1672:7
      %ret : Tensor = prim::If(%17) # /home/bowbao/repos/pytorch/torch/nn/functional.py:1672:4
        block0():
          %ret.1 : Tensor = aten::addmm(%self.linear.bias, %8, %27, %12, %12) # /home/bowbao/repos/pytorch/torch/nn/functional.py:1674:14
          -> (%ret.1)
        block1():
          %output.1 : Tensor = aten::matmul(%8, %27) # /home/bowbao/repos/pytorch/torch/nn/functional.py:1676:17
          %output.3 : Tensor = aten::add_(%output.1, %self.linear.bias, %12) # /home/bowbao/repos/pytorch/torch/nn/functional.py:1678:12
          -> (%output.3)
      %10 : Tensor[] = aten::append(%res.1, %ret) # repro_append.py:11:12
      -> (%2)
  %11 : Tensor = aten::cat(%res.1, %3) # repro_append.py:12:15
  return (%11)
"""
```

**Comment by eellison** [comment]
Loop outputs are variables that are assigned to in the loop.`res.append(self.linear(x[i]))` does not assign to a variable in the environment, so it's not an output of the loop. 

The `remove mutation` pass only removes mutation where it's semantically valid to do so. One way to think of it is what IR graph would you expect this graph to have post remove_mutation. In this case there's no way to remove the mutation of the graph you printed while preserving the semantics (appending an arbitrary number of times to a list).

**Comment by BowenBao** [comment]
No variable being assigned due to that it is updating `res` instead (in the ir). Isn't that against SSA?

In that case what is the suggested step to handle these scenarios? Is there any pass that is going to deal with the inplace/aliasing issue here?

**Comment by eellison** [comment]
SSA in TorchScript doesn't mean that the values are immutable. Mutation is tracked separately from SSA. Here, the object pointer for `res` remains the same throughout the duration of the program, and so has a single SSA value, but gets mutated with append.

cc @suo, I looked at https://github.com/pytorch/pytorch/blob/16e3df3ac6c7a47d758543e30161eef9ca1a896f/torch/csrc/jit/OVERVIEW.md#handling-mutability but I didn't really see anything all that comprehensive that would answer this question. 

There isn't really a pass that could deal with the aten::append issue here bc there's no way to represent it without append (equivalently, there's no way to write this in ONNX). 

The `aten::add_` could be safely removed from looking at the Graph, but due to the way we represent values that exit the graph it fails to rremoved currently. @suo, this is an example of where context-sensitive heap analysis would help. I would say it's a non-goal for us to have *perfect* alias analysis, however we try to make it better when we see repeated limitations bc of it.

**Comment by spandantiwari** [comment]
>> SSA in TorchScript doesn't mean that the values are immutable. Mutation is tracked separately from SSA. Here, the object pointer for res remains the same throughout the duration of the program, and so has a single SSA value, but gets mutated with append.

@eellison - just looking at the graph, I would say that the mutation of `res.1` (that happens in the loop) is not being captured in the graph. Would you agree with that? 
Given that, I would like to clarify upfront - do you agree that there's a problem/gap here, especially for a downstream component such as ONNX exporter, when translating this graph to express the original intent of the program?

In case you don't see this as an issue, I would like to understand that better. If you agree that there's some gap here, then we should talk about how this can be fixed. 

>>>>There isn't really a pass that could deal with the aten::append issue here bc there's no way to represent it without append (equivalently, there's no way to write this in ONNX).

This is quite concerning because this is a pretty common usage in Python, and shows up in real models, such as seq2seq models using beamsearch. 

cc: @suo and @bzinodev for their inputs as well.

**Comment by eellison** [comment]
I think one additional source of confusion is the `aten::append` schema is wrong; it should be returning None. I opened up a [PR](https://github.com/pytorch/pytorch/pull/39536) to address this. 
```
print([].append(1) is None)
# None
```

So the IR printout should be
```
%10 : None = aten::append(%res.1, %ret) # repro_append.py:11:12
```

> I would say that the mutation of res.1 (that happens in the loop) is not being captured in the graph. Would you agree with that?

No, I would not. If you dump the alias analysis you'll see that `aten::append` is understood to be mutating res.1. 
```
graph.dump_alias_db()
...
===3. Writes===
%10 : Tensor[] = aten::append(%res.1, %ret) # test/elias.py:10:12
  %res.1, 
%output.3 : Tensor = aten::add_(%output.1, %self.linear.bias, %12) 
  WILDCARD for type Tensor, %output.1, 
...
```

I also opened up an [issue](https://github.com/pytorch/pytorch/issues/39539) to be more explicit about mutation in our IR printout. `append` would be printed as 
```
%10 : None = aten::append(%res.1 (!), %ret) # repro_append.py:11:12
```

I understand there's a gap here, but I would expect there to be a gap when you try to take an IR with mutation to a more limited, functional IR. I don't expect all pytorch models to work with ONNX.

**Comment by spandantiwari** [comment]
>> I think one additional source of confusion is the aten::append schema is wrong; it should be returning None. I opened up a PR to address this.

Yes, this bug really throws one off when reading the graph. Thanks for the PR.

>>No, I would not. If you dump the alias analysis you'll see that aten::append is understood to be mutating res.1.

For clarity, can you point towards exactly what in the dump shows that res.1 is being mutated? 

>>I also opened up an issue to be more explicit about mutation in our IR printout. append would be printed as

This is a good step. In addition to being explicit in the IR printout, it will be good to have this captured, maybe as "metadata" of sorts, in the graph itself so that other passes could possibly leverage this info.

>>I understand there's a gap here, but I would expect there to be a gap when you try to take an IR with mutation to a more limited, functional IR. I don't expect all pytorch models to work with ONNX.

The TorchScript backend interface doc mentions the new functionalization pass. Can the functionalization pass currently not address this issue and produce a functional IR graph?

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
